Late-escaping of output, removal of $_REQUEST

stevegrunwell opened the pull request:
We're looking to use this on a platform that champions "escape all the things!", so I've gone through and added late-escaping wherever it was missing, following the [WordPress VIP standards](https://vip.wordpress.com/documentation/code-review-what-we-look-for/) (as they tend to know a thing or two about securing WordPress 😉).

This PR also replaces instances of `$_REQUEST` with `$_POST` or `$_GET`, based on how the data is being received.

Thank you for putting this together!

stevegrunwell commented:
Hmm, seem to be encountering an infinite login loop when I hit mysite.com/wp-login.php directly; will troubleshoot and have a patch shortly, as it may be connected to my `$_REQUEST` adjustments.

stevegrunwell commented:
That's exactly what it was, the 'redirect_to' value was getting lost, resulting in an infinite login loop.

georgestephanis commented on providers/class.two-factor-email.php:
```diff
@@ -64,11 +64,11 @@ function authentication_page( $user ) {
 
 	function validate_authentication( $user ) {
-		return $this->validate_token( $user->ID, $_REQUEST['two-factor-email-code'] );
+		return $this->validate_token( $user->ID, $_POST['two-factor-email-code'] );
```
This I had intentionally left as $_REQUEST in case we later modify the emails to include a link they can follow with the $_GET argument pre-filled.

georgestephanis commented on providers/class.two-factor-email.php:
```diff
@@ -64,11 +64,11 @@ function authentication_page( $user ) {
 		</script>
 		<?php
-		submit_button( __( 'Log In', 'two-factor' ) );
+		submit_button( esc_html__( 'Log In', 'two-factor' ) );
```
You're escaping this too early.  The string should be passed in to `submit_button()` unescaped, as it will escape it in a contextually correct way when printing.

https://github.com/WordPress/WordPress/blob/c6cff377e7aa9cf566a7511e54a470671cf5088a/wp-admin/includes/template.php#L1956

georgestephanis commented on providers/class.two-factor-email.php:
```diff
@@ -51,7 +51,7 @@ function authentication_page( $user ) {
 		<p>
 			<label for="authcode"><?php esc_html_e( 'Verification Code:' ); ?></label>
-			<input type="tel" name="two-factor-email-code" id="authcode" class="input" value="" size="20" pattern="[0-9]*" />
+			<input type="text" name="two-factor-email-code" id="authcode" class="input" value="" size="20" pattern="[0-9]*" />
```
Leaving this as `tel` results in a numpad input on mobile.  If you're going to change the input type, please change to an input type that keeps this ux.

stevegrunwell replied:
Using `input[type="text"]` with the numeric pattern [should be enough to trigger the numeric keypad](http://bradfrost.com/blog/mobile/better-numerical-inputs-for-mobile-forms/) (at the bottom of the article Brad points out the pattern attribute is enough to trigger the behavior), at least on iOS, though it would be beneficial to [do further device testing using the "Text pattern=[0-9]" input here](http://bradfrost.com/demo/tel/).

Screenshot of a text input with the pattern attribute in Xcode's iOS simulator (iPhone 6, iOS 8.2):

![ios simulator screen shot jul 13 2015 5 27 45 pm](https://cloud.githubusercontent.com/assets/233836/8661359/074de6fc-2985-11e5-8b10-b92a26f8c0fc.png)

georgestephanis commented on providers/class.two-factor-email.php:
```diff
@@ -39,8 +39,8 @@ function delete_token( $user_id ) {
-		$subject = sprintf( __( 'Your login confirmation code for %s', 'two-factor' ), get_bloginfo( 'name' ) );
-		$message = sprintf( __( 'Enter %s to log in.', 'two-factor' ), $token );
+		$subject = sprintf( esc_html__( 'Your login confirmation code for %s', 'two-factor' ), get_bloginfo( 'name' ) );
+		$message = sprintf( esc_html__( 'Enter %s to log in.', 'two-factor' ), $token );
```
`wp_mail()` by default does not use an HTML content type, it sends as plain text.  As such, escaping it with an HTML context would be inaccurate and potentially break things.

georgestephanis commented on providers/class.two-factor-email.php:
```diff
@@ -39,8 +39,8 @@ function delete_token( $user_id ) {
 
-		$subject = sprintf( __( 'Your login confirmation code for %s', 'two-factor' ), get_bloginfo( 'name' ) );
-		$message = sprintf( __( 'Enter %s to log in.', 'two-factor' ), $token );
+		$subject = sprintf( esc_html__( 'Your login confirmation code for %s', 'two-factor' ), get_bloginfo( 'name' ) );
```
Subject lines don't normally support html, so content shouldn't be escaped as such?

georgestephanis commented on providers/class.two-factor-dummy.php:
```diff
@@ -20,7 +20,7 @@ function authentication_page( $user ) {
 		<p><?php esc_html_e( 'Are you really you?', 'two-factor' ); ?></p>
 		<?php
-		submit_button( __( 'Yup.', 'two-factor' ) );
+		submit_button( esc_html__( 'Yup.', 'two-factor' ) );
```
As noted below -- https://github.com/georgestephanis/two-factor/pull/4/files#r34511775 -- this shouldn't be escaped here.

georgestephanis commented on class.application-passwords-list-table.php:
```diff
@@ -8,10 +8,10 @@ class Application_Passwords_List_Table extends WP_List_Table {
+			'name'      => esc_html__( 'Name', 'two-factor' ),
+			'created'   => esc_html__( 'Created', 'two-factor' ),
+			'last_used' => esc_html__( 'Last Used', 'two-factor' ),
+			'last_ip'   => esc_html__( 'Last IP', 'two-factor' ),
```
These columns probably shouldn't be escaped here, as you're providing the content, that the list table class can display in many different methods -- for example, as an attribute like so:

https://github.com/WordPress/WordPress/blob/b72f822a69ad634488b5a57aea99f0a78e42f777/wp-admin/includes/class-wp-list-table.php#L1213

Maybe calling `wp_strip_all_tags()` on it, or `wp_kses()` -- but escaping without knowing the usage context here feels wrong.

stevegrunwell replied:
Ah, good call.

georgestephanis commented on class.application-passwords-list-table.php:
```diff
@@ -71,7 +71,7 @@ protected function display_tablenav( $which ) {
 
 	public function single_row( $item ) {
-		echo '<tr data-slug="' . Application_Passwords::password_unique_slug( $item ) . '">';
+		echo '<tr data-slug="' . esc_attr( Application_Passwords::password_unique_slug( $item ) ) . '">';
```
Good catch.  I don't know if it will ever be useful in practice, as the content will only ever be a substr of a md5 hash, but it's still a good practice to be in.

georgestephanis commented on class.application-passwords.php:
```diff
@@ -86,7 +86,11 @@ public static function show_user_profile( $user ) {
+				<?php printf(
+					esc_html_x( 'Your new password for %1$s is %2$s.', 'application, password', 'two-factor' ),
+					'<strong>' . esc_html( $new_password_name ) . '</strong>',
+					'<kbd>' . self::chunk_password( $new_password ) . '</kbd>'
```
Since we're doing it anyways, should we also pass an `esc_html()` around the result in this row?

georgestephanis commented on class.application-passwords.php:
```diff
@@ -119,8 +123,8 @@ public static function catch_submission( $user_id ) {
 		$user_id = get_current_user_id();
-		if ( ! empty( $_REQUEST['delete_application_password'] ) ) {
-			$slug = $_REQUEST['delete_application_password'];
+		if ( ! empty( $_GET['delete_application_password'] ) ) {
```
Depending on how the AJAX processing endpoints are written, this may be better left as $_REQUEST to let one endpoint do both.  Maybe.  I'd need to see it built out first.

georgestephanis commented on class.application-passwords.php:
```diff
@@ -172,7 +176,7 @@ public static function delete_link( $item ) {
 		$delete_link = add_query_arg( 'delete_application_password', $slug );
 		$delete_link = wp_nonce_url( $delete_link, "delete_application_password-{$slug}", '_nonce_delete_application_password' );
-		return sprintf( '<a href="%1$s">%2$s</a>', esc_url( $delete_link ), esc_html__( 'Delete', 'two-factor' ) );
+		return sprintf( '<a href="%s">%s</a>', esc_url( $delete_link ), esc_html__( 'Delete', 'two-factor' ) );
```
No, the numbering was intentional.  Should the structure ever get changed around, it helps to make the sequence of arguments obvious.

georgestephanis commented on class.two-factor-core.php:
```diff
@@ -122,10 +122,10 @@ function show_two_factor_login( $user ) {
 		}
 
-		$redirect_to = isset( $_REQUEST['redirect_to'] ) ? $_REQUEST['redirect_to'] : $_SERVER['REQUEST_URI'];
+		$redirect_to = isset( $_POST['redirect_to'] ) ? $_POST['redirect_to'] : $_SERVER['REQUEST_URI'];
```
Leave this as $_REQUEST -- it's necessary if switching to a different provider on a branch I'm working on.

georgestephanis commented on class.two-factor-core.php:
```diff
@@ -134,7 +134,7 @@ function login_html( $user, $login_nonce, $redirect_to, $error_msg = '', $login_
 
 		$rememberme = 0;
-		if ( isset ( $_REQUEST[ 'rememberme' ] ) && $_REQUEST[ 'rememberme' ] ) {
+		if ( isset ( $_POST[ 'rememberme' ] ) && $_POST[ 'rememberme' ] ) {
```
Leave this as $_REQUEST -- it's necessary if switching to a different provider on a branch I'm working on.

georgestephanis commented on class.two-factor-core.php:
```diff
@@ -148,14 +148,14 @@ function login_html( $user, $login_nonce, $redirect_to, $error_msg = '', $login_
 		</form>
 
-		<p id="backtoblog"><a href="<?php echo esc_url( home_url( '/' ) ); ?>" title="<?php esc_attr_e( 'Are you lost?' ); ?>"><?php printf( __( '&larr; Back to %s' ), get_bloginfo( 'title', 'display' ) ); ?></a></p>
+		<p id="backtoblog"><a href="<?php echo esc_url( home_url( '/' ) ); ?>" title="<?php esc_attr_e( 'Are you lost?' ); ?>"><?php printf( esc_html__( '&larr; Back to %s', 'two-factor' ), get_bloginfo( 'title', 'display' ) ); ?></a></p>
```
This was intentionally using the core string.

https://github.com/WordPress/WordPress/blob/04793dce341e474dc16429150ee2a318df1ad305/wp-login.php#L219

georgestephanis commented:
I skipped a few remarks at the end between $_REQUEST and $_POST -- but I'd prefer if for the moment anyways, we could leave all of those as $_REQUEST until I get the backup methods branch sorted.

stevegrunwell commented:
@georgestephanis I've gone through and addressed your feedback on the PR, thank you for being so thorough! I've also reverted the `$_REQUEST`-related changes, but opened #6 so it doesn't get lost in the historical shuffle.

georgestephanis commented on class.two-factor-core.php:
```diff
@@ -217,7 +217,7 @@ function login_form_twostep() {
 			}
 
-			$this->login_html( $user, $login_nonce, $_REQUEST['redirect_to'], __( 'ERROR: Invalid verification code.', 'two-factor' ) );
+			$this->login_html( $user, $login_nonce, $_REQUEST['redirect_to'], esc_html__( 'ERROR: Invalid verification code.', 'two-factor' ) );
```
This is escaped inside `login_html()` here: https://github.com/georgestephanis/two-factor/blob/41bed06cc04ec0683ce34ac826cc8a29dd2e5790/class.two-factor-core.php#L144 -- doesn't need double escaping.

stevegrunwell commented:
Can you tell I've been burned by under-escaping before? ;)

georgestephanis commented:
:+1:

valendesigns commented:
I'm not sure if this PR is still needed anymore. There are aspects of it that made it into Master already. @stevegrunwell would you like to fix the merge conflicts so we can see what's different, or should we close the PR?

stevegrunwell commented:
@valendesigns It looks like the only thing left in the PR that hasn't been merged is converting the two-factor-input-code input from `type="tel"` to `type="text"`, as George and I [discussed earlier in the thread](#diff-for-comment-34511873). I'd say it's safe to close the PR :)

stevenkword commented:
We'll want to make sure to include the backup codes provider if we take another look at `$_REQUEST`.
